File: archived/ptune-log/src/core/models/tasks/MyTask.ts
// File: src/core/models/tasks/MyTask.ts
import { MyTaskFactory } from './MyTaskFactory';
import { PomodoroInfo } from './MyTask/PomodoroInfo';
import { GoogleTaskWrite } from './google/GoogleTaskWrite';
import { ReviewFlag } from './MyTask/ReviewFlag';
import { ReviewFlagNotesCodec } from './MyTask/ReviewFlagNotesCodec';

export class MyTask {
  constructor(
    public id: string,
    public title: string,
    public tasklist_id?: string,
    public note?: string,
    public parent?: string,
    public position?: string,
    public pomodoro?: PomodoroInfo,
    public status: 'needsAction' | 'completed' = 'needsAction',
    public subTasks?: MyTask[],
    public due?: string,
    public completed?: string, // UTC文字列
    public updated?: string,
    public started?: string, // UTC文字列
    public deleted: boolean = false,
    public reviewFlags?: Set<ReviewFlag>,
  ) {}

  toString(): string {
    const statusMarker = this.status === 'completed' ? '[x]' : '[ ]';
    const pomodoroStr = this.pomodoro ? ` ${this.pomodoro.toString()}` : '';
    const timeStr = this.formatTimeRange(this.started, this.completed);
    return `${statusMarker} ${this.title}${pomodoroStr}${timeStr}`;
  }

  private convertUtc(date: string): string {
    return date + (date.endsWith('Z') ? '' : 'Z');
  }

  private formatTimeRange(start?: string, end?: string): string {
    if (!start && !end) return '';
    const toTimeStr = (d: Date) =>
      `${d.getHours()}:${d.getMinutes().toString().padStart(2, '0')}:${d
        .getSeconds()
        .toString()
        .padStart(2, '0')}`;
    try {
      const parts = [];
      if (start) parts.push(toTimeStr(new Date(this.convertUtc(start))));
      if (end) parts.push(toTimeStr(new Date(this.convertUtc(end))));
      return ` ${parts.join(' - ')}`;
    } catch {
      return '';
    }
  }

  toGoogleTaskWrite(): GoogleTaskWrite {
    const notes: string[] = [];

    // ユーザメモ
    if (this.note) {
      notes.push(this.note);
    }

    // reviewFlags → notes
    if (this.note && this.note.trim().length > 0) {
      notes.push(ReviewFlagNotesCodec.strip(this.note) ?? this.note);
    }

    if (this.reviewFlags && this.reviewFlags.size > 0) {
      notes.push(ReviewFlagNotesCodec.encode(this.reviewFlags));
    }

    // pomodoro
    if (this.pomodoro) {
      let tomato = `🍅x${this.pomodoro.planned}`;
      if (this.pomodoro.actual !== undefined) {
        tomato += ` ✅x${this.pomodoro.actual}`;
      }
      notes.push(tomato);
    }

    const body: GoogleTaskWrite = {
      title: this.title,
      notes: notes.filter(Boolean).join(' ').trim(),
      status: this.status ?? 'needsAction',
    };

    if (this.id) body.id = this.id;
    if (this.parent) body.parent = this.parent;
    if (this.due) body.due = this.due;

    return body;
  }

  cloneWithoutActuals(): MyTask {
    const newTask = new MyTask(this.id, this.title);
    MyTaskFactory.copyTaskData(this, newTask);
    newTask.started = undefined;
    newTask.completed = undefined;
    if (newTask.pomodoro) {
      newTask.pomodoro.actual = undefined;
    }
    return newTask;
  }
}

File: archived/PtuneSync/PtuneSync/Models/MyTask.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace PtuneSync.Models;

/// <summary>
/// Google Tasks および Pomodoro 記録を統合するタスクモデル
/// </summary>
public class MyTask
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("title")]
    public string Title { get; set; } = string.Empty;

    [JsonPropertyName("tasklist_id")]
    public string? TaskListId { get; set; }

    [JsonPropertyName("note")]
    public string? Note { get; set; }

    [JsonPropertyName("parent")]
    public string? Parent { get; set; }

    [JsonPropertyName("position")]
    public string? Position { get; set; }

    [JsonPropertyName("pomodoro")]
    public PomodoroInfo? Pomodoro { get; set; }

    [JsonPropertyName("reviewFlags")]
    public HashSet<string>? ReviewFlags { get; set; }
    
    [JsonPropertyName("status")]
    public string Status { get; set; } = "needsAction"; // or "completed"

    [JsonPropertyName("subTasks")]
    public List<MyTask>? SubTasks { get; set; }

    [JsonPropertyName("due")]
    public string? Due { get; set; }

    [JsonPropertyName("completed")]
    public string? Completed { get; set; }

    [JsonPropertyName("updated")]
    public string? Updated { get; set; }

    [JsonPropertyName("started")]
    public string? Started { get; set; }

    [JsonPropertyName("deleted")]
    public bool Deleted { get; set; } = false;

    public MyTask() { }

    public MyTask(string id, string title, PomodoroInfo? pomodoro = null, string status = "needsAction")
    {
        Id = id;
        Title = title;
        Pomodoro = pomodoro;
        Status = status;
    }

    public override string ToString()
    {
        var statusMarker = Status == "completed" ? "[x]" : "[ ]";
        var pomodoroStr = Pomodoro != null ? $" {Pomodoro}" : string.Empty;
        var timeStr = FormatTimeRange(Started, Completed);
        return $"{statusMarker} {Title}{pomodoroStr}{timeStr}";
    }

    private string ConvertUtc(string date)
    {
        return date.EndsWith("Z") ? date : date + "Z";
    }

    private string FormatTimeRange(string? start, string? end)
    {
        if (string.IsNullOrEmpty(start) && string.IsNullOrEmpty(end)) return string.Empty;

        string ToTimeStr(DateTime d) =>
            $"{d.Hour}:{d.Minute:D2}:{d.Second:D2}";

        try
        {
            var parts = new List<string>();
            if (!string.IsNullOrEmpty(start))
                parts.Add(ToTimeStr(DateTime.Parse(ConvertUtc(start)).ToLocalTime()));
            if (!string.IsNullOrEmpty(end))
                parts.Add(ToTimeStr(DateTime.Parse(ConvertUtc(end)).ToLocalTime()));
            return " " + string.Join(" - ", parts);
        }
        catch
        {
            return string.Empty;
        }
    }

    public Dictionary<string, object> ToApiData()
    {
        var notes = new List<string>();
        if (!string.IsNullOrWhiteSpace(Note))
            notes.Add(Note);

        if (Pomodoro != null)
        {
            var tomato = $"🍅x{Pomodoro.Planned}";
            if (Pomodoro.Actual.HasValue)
                tomato += $" ✅x{Pomodoro.Actual}";
            notes.Add(tomato);
        }

        var body = new Dictionary<string, object>
        {
            ["title"] = Title,
            ["notes"] = string.Join(" ", notes).Trim(),
            ["status"] = Status
        };

        if (!string.IsNullOrEmpty(Id)) body["id"] = Id;
        if (!string.IsNullOrEmpty(Parent)) body["parent"] = Parent;
        if (!string.IsNullOrEmpty(Due)) body["due"] = Due;

        return body;
    }

    public MyTask CloneWithoutActuals()
    {
        var clone = (MyTask)MemberwiseClone();
        clone.Started = null;
        clone.Completed = null;
        if (clone.Pomodoro != null)
            clone.Pomodoro.Actual = null;
        return clone;
    }
}
